### PR TITLE
neovim: add 'autoconfigure' setting

### PR DESCRIPTION
**Pull request by teto**
certain plugins need a custom configuration (available in passthru.initLua) to work with nix. For instance unicode-vim needs to configure where to find the data:
```
vim.g.Unicode_data_directory="${self.unicode-vim}/autoload/unicode"
```
if true, the wrapper automatically appends those snippets when necessary

first setup in https://github.com/NixOS/nixpkgs/issues/352741

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by teto** [comment]
disabled by default to let users test first for instance with `neovim.overrideAttrs({ autoconfigure = true;})`. 
The only concern I have is with the naming. I would like us to address the autoaddition of plugin runtimeDeps in the neovim wrapper see https://github.com/NixOS/nixpkgs/issues/352738. This will start disabled by default as well. Is this going to be a different flag or the same for simplicity ?
Maybe something like that ?
```
wrapper = {
autoconfigure ? false
autoconfigureRuntimeDeps ? autoconfigure
...
}:
...
```

File: pkgs/applications/editors/neovim/wrapper.nix [hunk 1]
```diff
@@ -87,11 +91,19 @@ let
     packpathDirs.myNeovimPackages = myVimPackage;
     finalPackdir = neovimUtils.packDir packpathDirs;
 
+    luaPluginRC = let
+      op = acc: normalizedPlugin:
+           acc ++ lib.optional (finalAttrs.autoconfigure && normalizedPlugin.plugin.passthru ? initLua) normalizedPlugin.plugin.passthru.initLua;
+      in
+        lib.foldl' op [] pluginsNormalized;
```
Comment (mrcjkb) [inline pkgs/applications/editors/neovim/wrapper.nix:98]: question: `foldl'` will reverse the order (correct me if mistaken). Do we care about this?

Comment (teto) [reply pkgs/applications/editors/neovim/wrapper.nix:98]: I dont think so, the accumulator comes first.
applying
```
diff --git a/pkgs/applications/editors/neovim/tests/default.nix b/pkgs/applications/editors/neovim/tests/default.nix
index fb89d47aceb0..48b00d42f272 100644
--- a/pkgs/applications/editors/neovim/tests/default.nix
+++ b/pkgs/applications/editors/neovim/tests/default.nix
@@ -170,7 +170,10 @@ in
   '';
 
   nvim_with_autoconfigure = pkgs.neovim.overrideAttrs(oa: {
-    plugins = [ vimPlugins.unicode-vim ];
+    plugins = [
+      vimPlugins.unicode-vim
+      vimPlugins.fzf-hoogle-vim
+    ];
     autoconfigure = true;
     # legacy wrapper sets it to false
     wrapRc = true;
```
generated 
```
vim.cmd.source "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-init.vim"
vim.g.Unicode_data_directory="/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vimplugin-unicode.vim-2024-07-23/autoload/unicode"
vim.g.hoogle_fzf_cache_file = vim.fn.stdpath('cache')..'/hoogle_cache.json
```

**Review by mrcjkb — APPROVED** [review]

**Review by PerchunPak — APPROVED** [review]
LGTM

File: pkgs/applications/editors/neovim/tests/default.nix [hunk 2]
```diff
@@ -80,7 +80,12 @@ let
     sha256 = "1ykcvyx82nhdq167kbnpgwkgjib8ii7c92y3427v986n2s5lsskc";
   };
 
-  # neovim-drv must be a wrapped neovim
+  /* neovim-drv must be a wrapped neovim
+    - exposes lua config in $luarcGeneric
+    - exposes vim config in $vimrcGeneric
+
```
Comment (khaneliman) [inline pkgs/applications/editors/neovim/tests/default.nix:86]:
```suggestion
```

**Review by khaneliman — APPROVED** [review]
LGTM

**Review by GaetanLepage — APPROVED** [review]

File: pkgs/applications/editors/neovim/tests/default.nix (hunk 2 above)
Comment (teto) [reply pkgs/applications/editors/neovim/tests/default.nix:86]: I'll do this on next update ty

**Pull request merged**